src/odemis/acq/move.py

```diff
@@ -614,6 +805,237 @@ def _doCryoSwitchSamplePosition(self, future, target):
+        # For the rotation case: pos_cor calibration data is multipled by 2x due to historical reasons
+        # it is the radius of rotation -> we need the diameter, therefore 2x
+        # TODO: remove the 2x multiplication when the calibration data is updated
+        if has_rz:
+            transformed_pos["x"] = 2 * pos_cor[0] - pos["x"]
+            transformed_pos["y"] = 2 * pos_cor[1] - pos["y"]
+        else:
+            transformed_pos["x"] = pos["x"] + pos_cor[0]
+            transformed_pos["y"] = pos["y"] + pos_cor[1]
+
+        transformed_pos.update(fm_pos_active)
+
+        if fix_fm_plane:
+            # In FM mode, observe the features on sample on a fixed imaging plane
+            transformed_pos_fav = self.get_stage_pos_at_fav(transformed_pos, self.fav_pos_active)
+
+            if transformed_pos_fav is not None:
+                transformed_pos = transformed_pos_fav
+            # check if the transformed position is within the FM imaging range
+            if not isInRange(transformed_pos, stage_md[model.MD_FM_IMAGING_RANGE], {'x', 'y'}):
+                # only log warning, because transforms are used to get current position too
+                logging.warning(f"Transformed position {transformed_pos} is outside FM imaging range")
+
+        return transformed_pos
+
+    # Note: this transformation also consists of translation and rotation.
+    # The translation is along x and y axes. They are calculated based on
+    # the current position and correction parameters which are calibrated every session.
+    # The rotation angles are 180 degree around rz axis, and a rotation angle
+    # around rx axis which should also be calibrated at the beginning of the run.
+    # The rx angle is actually the same as the milling angle.
+    def _transformFromMeteorToSEM(self, pos: Dict[str, float]) -> Dict[str, float]:
+        """
+        Transforms the current stage position from the meteor/FM imaging area
+        to the SEM imaging area.
+        :param pos: (dict str->float) the initial stage position.
+        :return: (dict str->float) the transformed stage position.
+        """
+        stage_md = self.stage.getMetadata()
+        pos_cor = stage_md[model.MD_POS_COR]
+        sem_pos_active = stage_md[model.MD_FAV_SEM_POS_ACTIVE]
+
+        # check if the stage positions have rz axes
+        if not ("rz" in pos and "rz" in sem_pos_active):
+            raise ValueError(f"The stage position does not have rz axis. pos={pos}, sem_pos_active={sem_pos_active}")
+
+        # whether we need to rotate around the z axis (180deg)
+        has_rz = not isNearPosition(pos, sem_pos_active, {"rz"},
+                                    atol_rotation=ATOL_ROTATION_TRANSFORM)
+
+        # compensate the impact of fixing the fm imaging plane on the current stage position
+        # based on the value of original imaging plane value saved before fixing the fm imaging plane
+        new_pos = pos.copy()
```

Comment:
I think this is missing transformed_pos in the case there no stage_pos_before_fav_fm.  And I don't think we can assume it's always there